himbaechel/uarch/xilinx/pack_dsp_xc7.cc

```diff
@@ -0,0 +1,165 @@
+/*
+ *  nextpnr -- Next Generation Place and Route
+ *
+ *  Copyright (C) 2020  Myrtle Shah <gatecat@ds0.me>
+ *  Copyright (C) 2023  Hans Baier <hansfbaier@gmail.com>
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include <boost/algorithm/string.hpp>
+#include "pack.h"
+
+#define HIMBAECHEL_CONSTIDS "uarch/xilinx/constids.inc"
+#include "himbaechel_constids.h"
+
+NEXTPNR_NAMESPACE_BEGIN
+
+void XC7Packer::walk_dsp(CellInfo *root, CellInfo *current_cell, int constr_z)
+{
+    CellInfo *cascaded_cell = nullptr;
+
+    auto check_illegal_fanout = [&] (NetInfo *ni, std::string port) {
+        if (ni->users.entries() > 1)
+            log_error("Port %s connected to net %s has more than one user", port.c_str(), ni->name.c_str(ctx));
+
+        PortRef& user = *ni->users.begin();
+        if (user.cell->type != id_DSP48E1_DSP48E1)
+            log_error("User %s of net %s is not a DSP block, but %s",
+                user.cell->name.c_str(ctx), ni->name.c_str(ctx), user.cell->type.c_str(ctx));
+    };
+
+    // see if any cascade outputs are connected
+    for (auto port : current_cell->ports) {
+        if (!boost::contains(port.first.str(ctx), "COUT")) continue;
+        NetInfo *cout_net = port.second.net;
+
+        if (cout_net == nullptr) continue;
```

Comment:
Sorry, I think my previous comment got misunderstood - this is the only place where the empty check is needed, it should be removed down below